File: gender/python/gender-robust-clean.py
#!/usr/bin/env python

host = '127.0.0.1'
port = 9199
name = 'test'

import jubatus
from jubatus.classifier.types import datum
import msgpackrpc
import time

class Retryable(object):
    def __init__(self, client_class, host, port, timeout = 10):
        self._client_class = client_class
        self._host = host
        self._port = port
        self._timeout = timeout
        self._client = None
        self.reopen()

    def reopen(self):
        if not self._client is None:
            self._client.get_client().close()
        self._client = self._client_class(self._host, self._port)
        self._client.get_client()._timeout = self._timeout

    def __getattr__(self, name):
        return getattr(self._client, name)

    def with_retry(self, retry_max, retry_interval):
        return Retryable.RetryFeature(self, retry_max, retry_interval)

    class RetryFeature(object):
        def __init__(self, client_holder, retry_max, retry_interval):
            self._holder = client_holder
            self._max = retry_max
            self._interval = retry_interval

        def __getattr__(self, name):
            return Retryable.Executer(self._holder, name, self._max, self._interval)

    class Executer(object):
        def __init__(self, holder, name, retry_max, retry_interval):
            self._holder = holder
            self._name = name
            self._max = retry_max
            self._interval = retry_interval

        def __call__(self, *args):
            retry_count = 0
            while True:
                try:
                    body = getattr(self._holder._client, self._name)
                    return body(*args)

                except (msgpackrpc.error.TransportError, msgpackrpc.error.TimeoutError) as e:
                    retry_count += 1
                    if retry_count >= self._max:
                        raise

                    self._holder.reopen()

                    print e
                    time.sleep(self._interval)
                    continue
        
client = Retryable(jubatus.Classifier, host, port, 1)

train_data = [
    ('male',   datum([('hair', 'short'), ('top', 'sweater'), ('bottom', 'jeans')], [('height', 1.70)])),
    ('female', datum([('hair', 'long'),  ('top', 'shirt'),   ('bottom', 'skirt')], [('height', 1.56)])),
    ('male',   datum([('hair', 'short'), ('top', 'jacket'),  ('bottom', 'chino')], [('height', 1.65)])),
    ('female', datum([('hair', 'short'), ('top', 'T shirt'), ('bottom', 'jeans')], [('height', 1.72)])),
    ('male',   datum([('hair', 'long'),  ('top', 'T shirt'), ('bottom', 'jeans')], [('height', 1.82)])),
    ('female', datum([('hair', 'long'),  ('top', 'jacket'),  ('bottom', 'skirt')], [('height', 1.43)])),
#    ('male',   datum([('hair', 'short'), ('top', 'jacket'),  ('bottom', 'jeans')], [('height', 1.76)])),
#    ('female', datum([('hair', 'long'),  ('top', 'sweater'), ('bottom', 'skirt')], [('height', 1.52)])),
    ]

client.with_retry(5, 1).train(name, train_data)

raw_input('now, classify: ')

test_data = [
    datum([('hair', 'short'), ('top', 'T shirt'), ('bottom', 'jeans')], [('height', 1.81)]),
    datum([('hair', 'long'),  ('top', 'shirt'),   ('bottom', 'skirt')], [('height', 1.50)]),
]

results = client.with_retry(5, 1).classify(name, test_data)

for result in results:
    for r in result:
        print r.label, r.score
    print

File: gender/ruby/gender-robust-clean.rb
#!/usr/bin/env ruby

require 'jubatus/classifier/client'
require 'jubatus/classifier/types'

host = "127.0.0.1"
port = 9199
name = "test"

RETRY_MAX = 5
RETRY_INTERVAL = 3.0

def with_retry(client, max = RETRY_MAX, interval = RETRY_INTERVAL)
  retry_count = 0
  begin
    yield(client)

  rescue MessagePack::RPC::TimeoutError, MessagePack::RPC::TransportError => e
    raise if (retry_count += 1) >= max

    client.get_client.close
    $stderr.puts e
    sleep interval
    retry
  end
end

if !ARGV.empty?
  host_spec = ARGV[0]
  if host_spec =~ /^(.+):(\d+)$/
    host = $1
    port = $2.to_i
  else
    host = $1
  end
end
$stderr.puts "connect server #{host}:#{port}"

client = Jubatus::Classifier::Client::Classifier.new(host, port)

train_data =
  [
   ["male",   Jubatus::Classifier::Datum.new([["hair", "short"], ["top", "sweater"], ["bottom", "jeans"]], [["height", 1.70]])],
   ["female", Jubatus::Classifier::Datum.new([["hair", "long"],  ["top", "shirt"],   ["bottom", "skirt"]], [["height", 1.56]])],
   ["male",   Jubatus::Classifier::Datum.new([["hair", "short"], ["top", "jacket"],  ["bottom", "chino"]], [["height", 1.65]])],
   ["female", Jubatus::Classifier::Datum.new([["hair", "short"], ["top", "T shirt"], ["bottom", "jeans"]], [["height", 1.72]])],
   ["male",   Jubatus::Classifier::Datum.new([["hair", "long"],  ["top", "T shirt"], ["bottom", "jeans"]], [["height", 1.82]])],
   ["female", Jubatus::Classifier::Datum.new([["hair", "long"],  ["top", "jacket"],  ["bottom", "skirt"]], [["height", 1.43]])],
#   ["male",   Jubatus::Classifier::Datum.new([["hair", "short"], ["top", "jacket"],  ["bottom", "jeans"]], [["height", 1.76]])],
#   ["female", Jubatus::Classifier::Datum.new([["hair", "long"],  ["top", "sweater"], ["bottom", "skirt"]], [["height", 1.52]])],
  ]

with_retry(client) { |c| c.train(name, train_data) }

$stdout.write "now, classify: "
$stdout.flush
$stdin.gets

test_data =
  [
   Jubatus::Classifier::Datum.new([["hair", "short"], ["top", "T shirt"], ["bottom", "jeans"]], [["height", 1.81]]),
   Jubatus::Classifier::Datum.new([["hair", "long"],  ["top", "shirt"],   ["bottom", "skirt"]], [["height", 1.50]]),
  ]

results = with_retry(client) { |c| c.classify(name, test_data) }

results.each { |result|
  result.each { |(label, score)|
    puts(label + " " + score.to_s)
  }
  puts
}
